Issue8758 migrate nRF sw pwm config to dt

gamnes opened the pull request:
This PR moves SW PWM configuration from Kconfig to DT. 
It also adds some new configuration options in DT, and updates the nRF SW PWM driver to use these.
Some samples had to be updated to be compatible with the new DT setup.

gamnes commented:
A challenge presented itself because we have _some configuration in Kconfig, and __some in Device Tree. 

For example, for the fade_led example, the nrf52_pca10040 board is capable of running both SW PWM and HW PWM. 
The choice between nRF SW or HW PWM is done by the user in Kconfig through: for SW: (CONFIG_PWM_NRF5_SW) vs for HW: (CONFIG_PWM_NRFX and CONFIG_PWM_0). 
The problem is that the configuration of SW and HW PWM driver is now done in DT overlay files, nrf51_pca10028.overlay, and nrf52_pca10040.overlay; and the nrf52 cannot (should not) include both the SW PWM driver (sw_pwm, status ="ok") and the HW PWM driver (&pwm0, status="ok"). 
Have a look in the code how I solved this by forcing nRf52 to use HW PWM in Kconfig, and writing a comment in the overlay how SW could be used if desired.

gamnes commented on boards/arm/bbc_microbit/bbc_microbit.dts:
```diff
@@ -7,6 +7,32 @@
 /dts-v1/;
 #include <nordic/nrf51822_qfaa.dtsi>
 
+/* External edge connector pin mappings to nRF51 GPIO pin numbers.
```
This was done because I wanted to get hold of the EXT_P0_GPIO_PIN define in DT board.overlay file. I do not expect a user to know that Microbit P0 connection is connected to P0.03 on the nRF51822 - and thus know that he should write ch0-pin = <3>; in his board.overlay file for the application. Using <EXT_P0_GPIO_PIN> made much more sense to me.

gamnes replied:
Defines are removed again in new push - they were not used anywhere else so no need to change this file as part of this PR (if we do want to change it, that is.

gamnes commented on drivers/pwm/Kconfig.nrf5_sw:
```diff
@@ -19,28 +19,9 @@ if PWM_NRF5_SW
 config PWM_NRF5_SW_0_DEV_NAME
 	string "Nordic Semiconductor nRF5x series S/W PWM Device Name"
-	default "PWM_0"
+	default "SW_PWM"
```
Changed to avoid confusion between HW PWM driver and SW PWM driver. Is there a mechanisms to handle multiple drivers being labeled the same thing?

SebastianBoe commented on dts/arm/nordic/nrf52840.dtsi:
```diff
@@ -60,6 +60,16 @@
+		status = "disabled";
+		label = "SW_PWM";
+	};
+
```
Is it possible/desirable to have a nrf.dtsi that declares sw_pwm: sw-pwm {} with default values and then have individual SoC's override the values that they need to set differently?

gamnes replied:
Yes, that is possible. 
Steps: 
- add new file nrf.dtsi in same folder as nrf52832.dtsi. 
- Define sw_pwm: ... node in this file instead of nrf52832.dtsi
-- Leave out timer-instance, clock-prescaler, ppi-base, and gpiote-base.
- In nrf52832.dtsi, include modify this node by writing &sw_pwm { ... and then set timer-instance, clock-prescaler, ppi-base, and gpiote-base here.
- Works the same from application - no changes there. 

If this is desirable, i don't know?

anangl replied:
Yes, I think it's a good idea to have a common definition of the node and override only the parts that need be changed for particular SoCs.

gamnes replied:
Done. In the common nrf5.dtsi, i set timer-instance, clock-prescaler, ppi-base, gpiote-base all to 0. If we want to keep this as the default config for nrf52 instead that can be done, but i find it "cleaner" if both nrf51 and nrf52 actually change these values to what they should be

SebastianBoe commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -244,16 +244,15 @@ static int pwm_nrf5_sw_init(struct device *dev)
 	.timer = NRF_TIMER1,
-	.ppi_base = 7,
-#else
+#elif (DT_NORDIC_NRF_SW_PWM_SW_PWM_TIMER_INSTANCE == 2)
```
What about

`.timer = NRF_TIMER ## DT_NORDIC_NRF_SW_PWM_SW_PWM_TIMER_INSTANCE`

?

For easier porting to new platforms.

gamnes replied:
As you mentioned; #include <misc/util.h>, UTIL_CAT(NRF_TIMER, DT_NORDIC_NRF_SW_PWM_SW_PWM_TIMER_INSTANCE) will work for this purpose.
But, leaving as is for now unless someone objects

anangl replied:
I support @SebastianBoe's suggestion. This will be easier in maintenance.
`_CONCAT` seems to be a standard Zephyr's macro for this purpose.

gamnes replied:
_CONCAT works. Changed

lemrey commented:
I actually think we should be using the hardware PWM where we have it, instead of defaulting to the software implementation in every sample.

carlescufi commented:
@anangl can you please review this one?

gamnes commented:
> I think we should be using the hardware PWM where we have it, instead of defaulting to the software implementation.

Agreed, that would make sense. - But this should perhaps be done in a separate PR as its not related to Kconfig/DT migration (?)

anangl commented on dts/bindings/pwm/nordic,nrf-sw-pwm.yaml:
```diff
@@ -0,0 +1,59 @@
+
+    clock-prescaler:
+      type: int
+      description: Clock prescaler for timer used for generating the PWM output signals
```
A description of the prescaler value meaning would be nice. Something like "frequency = 16 MHz / 2^prescaler".

gamnes replied:
Done

anangl commented on samples/basic/servo_motor/prj_bbc_microbit.conf:
```diff
@@ -2,4 +2,3 @@ CONFIG_STDOUT_CONSOLE=y
 CONFIG_PRINTK=y
 CONFIG_PWM=y
 CONFIG_PWM_NRF5_SW=y
-CONFIG_PWM_NRF5_SW_0_CLOCK_PRESCALER=3
```
After this change, the whole file seems to be unneeded. `CONFIG_PWM_NRF5_SW` is `y` by default, when PWM is enabled.

gamnes replied:
You are correct. Will remove file

gamnes commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -65,15 +76,8 @@ static u8_t pwm_channel_map(struct pwm_data *data, u8_t map_size,
 		}
 	}
 
-	/* find a free entry */
```
Reverted in new push. Now we allow finding free PWM entry as before. SW PWM GPIOs are not set in DT for SW PWM.

gamnes commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -49,6 +49,16 @@
 
+	sw_pwm: sw-pwm {
+		compatible = "nordic,nrf-sw-pwm";
+		timer-instance = <1>;
```
Maybe something that can be added in the future.

gamnes commented on samples/basic/blink_led/nrf52840_pca10056.overlay:
```diff
@@ -0,0 +1,6 @@
+
+&sw_pwm {
+	status = "ok";
+	ch0-pin = <13>;
```
Maybe future

gamnes commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -12,6 +12,17 @@
+/**
+ * @brief Defines the number of PWM channels (outputs)
+ */
+#define PWM_0_MAP_SIZE 3
```
"Value" is now configurable via DTS - DTS allows you to set timer CC count. PWM map size will be set as this value minus 1 (to have room for the top period value).
Pin selection in DT was reverted. Driver now works as before - pins are chosen at run time and can be changed.

gamnes commented:
PR update with new content. Driver now works as before where SW PWM outputs are possible to change run-time. All this PR does it is move Kconfig options, as well as some of the static configuration that was done in the driver to DT, and then use these DT defines. Some cleanup of the driver source was also done.

I was tempted to remove the "int" settings in the default sw_pwm node of the nrf5.dtsi, but set them as 0 for now (easier to see what needs to be set when "inherited").

All samples i could find that relied on the nRF SW PWM driver has been updated to enable the SW_PWM node so that they should now work with the new DT config.

Please review and let me know if there are any issues. Thanks.

anangl commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -12,6 +12,9 @@
 LOG_MODULE_REGISTER(pwm_nrf5_sw);
 
+#define PWM_0_MAP_SIZE \
+	(DT_NORDIC_NRF_SW_PWM_SW_PWM_TIMER_CAPTURE_COMPARE_COUNT - 1)
```
Why `- 1`?

gamnes replied:
-1 because we need one compare register to set the period. So the last compare is used to "reset"

gamnes replied:
Not needed anymore since we changed the DT name to be CHANNEL_COUNT - hence the MAP_SIZE is equal to channel count (user just have to make sure there the channel count is one less than the number of capture compare registers) - as mentioned in yaml description.

anangl commented on dts/bindings/pwm/nordic,nrf-sw-pwm.yaml:
```diff
@@ -0,0 +1,65 @@
+      description: Human readable string describing the device (used by Zephyr for API name)
+      generation: define
+
+    ch0-pin:
```
This should be removed, as well as `ch1-pin` and `ch2-pin`.

gamnes replied:
Yes

anangl commented on dts/bindings/pwm/nordic,nrf-sw-pwm.yaml:
```diff
@@ -0,0 +1,65 @@
+      category: required
+      generation: define
+
+    timer-capture-compare-count:
```
This should be rather `channel-count`. The PWM user does not need to know how the driver is implemented, and that for each PWM channel it uses one CC register.

gamnes replied:
Changed to channel-count, made a comment that this needs to be one less than timer capture compare reigsters.

anangl commented on dts/arm/nordic/nrf5.dtsi:
```diff
@@ -0,0 +1,13 @@
+	sw_pwm: sw-pwm {
+		compatible = "nordic,nrf-sw-pwm";
+		timer-instance = <0>;
+		timer-capture-compare-count = <0>;
```
I don't see the point in specifying all properties here, especially with invalid values like this one. What's wrong in having only the common ones here (i.e. `compatible`, `status`, and `label`)?

gamnes replied:
Not needed - removed.

anangl commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -1,6 +1,7 @@
 #include <arm/armv6-m.dtsi>
 #include <dt-bindings/i2c/i2c.h>
 #include <dt-bindings/gpio/gpio.h>
+#include <nordic/nrf5.dtsi>
```
`#include "nrf5.dtsi"` would be more adequate. And I think the file should be named "nrf5_common.dtsi".

gamnes replied:
Was done to match nrf51822_qfaa.dtsi, etc, files. But using "nrf5.dtsi" now.

gamnes replied:
Changed name to nrf_common.dtsi, changed include to be as stated (just with the new name).

anangl commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -244,23 +246,20 @@ static int pwm_nrf5_sw_init(struct device *dev)
-#endif
-	.gpiote_base = 0,
+	.timer = _CONCAT(NRF_TIMER, DT_NORDIC_NRF_SW_PWM_SW_PWM_TIMER_INSTANCE),
+	.ppi_base = DT_NORDIC_NRF_SW_PWM_SW_PWM_PPI_BASE,
```
This is a single instance driver, so you can use `DT_NORDIC_NRF_SW_PWM_0_*` macros. They are more readable, and shorter.

gamnes replied:
Changed to use the new DT defines.

anangl commented on samples/basic/blink_led/nrf52832_mdk.overlay:
```diff
@@ -0,0 +1,5 @@
+
```
Why do you need this file? For boards with nRF52 the HW PWM driver is enabled by default.

gamnes replied:
SW-PWM setting "CONFIG_PWM_NRF5_SW_0_CLOCK_PRESCALER" was set and used for this example on the nrf52832_mdk board earlier. I did not want to change that, so I updated and added this file. Now that it has changed, no need to keep this overlay since HW PWM is used. 
File removed.

anangl commented on samples/basic/fade_led/prj_nrf52_pca10040.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_PRINTK=y
```
This file is no longer needed. All these settings will get such values also without it.

gamnes replied:
Part of demo how to use SW-PWM.
Removed.

anangl commented on samples/basic/fade_led/nrf52_pca10040.overlay:
```diff
@@ -0,0 +1,17 @@
+&pwm0 {
+	status = "ok";
+	ch0-pin = <17>;
```
This file is not needed. The `pwm0` node is configured properly at the board level.

gamnes replied:
Part of demonstration on how to use SW-PWM driver instead of HW.
Removed.

anangl commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -153,3 +154,11 @@
+	clock-prescaler = <0>;
+	ppi-base = <7>;
+	gpiote-base = <0>;
+};
```
I think `status = "ok";` should be added here. Then, there would be no need to add overlay files in most cases (only when for example `clock-prescaler` has to be modified).

gamnes replied:
Agreed.

anangl commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -153,3 +154,12 @@
+
+&sw_pwm {
+	timer-instance = <1>;
+	channel-count = <4>;
```
`<3>`
4 is the total number of compare channels in TIMER1.
Same applies to TIMER2 used for nRF52s.

gamnes replied:
Of course, changed.

anangl commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -12,6 +12,8 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(pwm_nrf5_sw);
 
+#define PWM_0_MAP_SIZE DT_NORDIC_NRF_SW_PWM_0_CHANNEL_COUNT
```
Maybe it would be good to check here if the number of channels specified in DTS is not too high?
Though the check would not look very nice:
```
/* One compare channel is needed to set the PWM period, hence +1. */
#if ((DT_NORDIC_NRF_SW_PWM_0_CHANNEL_COUNT + 1) > \
     (_CONCAT(_CONCAT(TIMER, DT_NORDIC_NRF_SW_PWM_0_TIMER_INSTANCE), _CC_NUM)))
```

gamnes replied:
I think this is a good idea - at least the user will be notified that his parameter is wrongly set, instead of just running with it and failing "silently".

Doesnt the check ....CHANNEL_COUNT < ...._CC_NUM look better? We just need to be less than number of CC, we dont care if the user isnt using all of them?
If we do +1, it needs to be equal instead of > i believe.

anangl replied:
> Doesnt the check ....CHANNEL_COUNT < ...._CC_NUM look better? We just need to be less than number of CC, we dont care if the user isnt using all of them?
> If we do +1, it needs to be equal instead of > i believe.

My intention was to place an `#error` inside this check.

gamnes replied:
Yes, that makes sense (Y)

gamnes replied:
Included. Will update PR soon

gamnes replied:
/* One compare channel is needed to set the PWM period, hence +1. */
#define PWM_0_MAP_SIZE DT_NORDIC_NRF_SW_PWM_0_CHANNEL_COUNT
#if ((DT_NORDIC_NRF_SW_PWM_0_CHANNEL_COUNT + 1) > \
	 (_CONCAT(_CONCAT(TIMER, DT_NORDIC_NRF_SW_PWM_0_TIMER_INSTANCE), _CC_NUM)))
#error "PWM channel-count must be one less than timer compare channels count."
#endif

anangl replied:
Actually, it doesn't have to be one less then CC count. It can be lower.
What about just "Invalid number of PWM channels configured."?

gamnes replied:
Agreed.

anangl commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -7,11 +7,22 @@
+     (_CONCAT( \
+        _CONCAT(TIMER, DT_NORDIC_NRF_SW_PWM_0_TIMER_INSTANCE), \
+        _CC_NUM)) \
+     )
```
Either indent this line one space less or move the parenthesis to the previous line (this I'd prefer, and I guess it will be more consistent with Zephyr coding style). Now it is more misleading than helping.

gamnes replied:
Moved up

anangl commented on drivers/pwm/pwm_nrf5_sw.c:
```diff
@@ -7,11 +7,21 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(pwm_nrf5_sw);
 
+/* One compare channel is needed to set the PWM period, hence +1. */
```
This comment refers to line [18](https://github.com/zephyrproject-rtos/zephyr/pull/12390/files#diff-0f7fc8bd683a80e71bb5665dfca79dfbR18) (`+ 1` occurs there).
I think the best would be to move the line with `#define` after the check (first validate, then use).

gamnes replied:
Agreed. moved.

anangl commented:
The [last correction](https://github.com/zephyrproject-rtos/zephyr/compare/f4c07f0614c0e4cdbb6775ae860d4b6c9c746b84..9fb83aa20043f43285614a23c499cc49ef928e2d) in indentation was unnecessary (actually, the code look worse now). There was only a warning reported for this line. Previously the problem was that you indented lines with 8 spaces instead of TAB.

gamnes committed:
dts: nrf: SW PWM device node added to nRF5 devices with yaml binding

1. SW PWM device node added to common nrf5_common.dtsi
2. SW PWM node set in all nRF5x DTSI files.
   Different initial settings for nRF51 and nRF52 devices.
   Status is ok by default for nRF51.
3. Added yaml binding for Nordic SW PWM node.
4. Set codeowner of nordic dts bindings to @anangl

Signed-off-by: Gaute Gamnes <gaute.gamnes@nordicsemi.no>

gamnes committed:
drivers: pwm: nrf: Migrate SW PWM Kconfig options to DT

1. Kconfig option Clock prescaler removed.
2. Modified pwm_nrf5_sw.c driver to use DT
   defines instead of Kconfig, and also use new
   DT options (timer, ppi/gpiote, etc).
3. Cleanup some code.

Signed-off-by: Gaute Gamnes <gaute.gamnes@nordicsemi.no>

gamnes committed:
samples: servo_motor: Fix servo_motor sample to use new DT options

1. Remove microbit Kconfig conf file.
2. Include microbit DT overlay with prescaler setting.
3. Set codeowner of servo_motor bbc_microbit to @jhe

Signed-off-by: Gaute Gamnes <gaute.gamnes@nordicsemi.no>
